robocopkaka/customer_ticketing_system
# Ft refactor controllers to use sessions table

#### What does this PR do?
Adds a session resource
#### Description of Task to be completed?
* Add session resource - controller, model, routes
* Replace `knock` login with custom login using session
* Add custom `authenticate` calls to replace the ones from `knock`
* Modify existing tests to use new session resource
#### How should this be manually tested?
a) To create a new session
Send:

```
curl -L -X POST 'http://localhost:3001/customers/login' \
-H 'Content-Type: application/json' \
--data-raw '{
	"email": "customer-Sid-3@gmail.com",
	"password": "password"
}'
```

Which should return:

```.json
{
    "data": {
        "session": {
            "id": "sess-733eceff3cb946eebabe039c688d823c",
            "active": true,
            "expires_at": "2020-06-04T17:10:16Z",
            "deleted": false,
            "deleted_at": null,
            "role": "Customer",
            "user_id": "cust-0694b4d4ebedf7b13a55b5734720f989",
            "user_agent": "PostmanRuntime/7.25.0"
        }
    },
    "extra": ""
}
```

b) i) To fetch a session
Send:

```
curl -L -X GET 'http://localhost:3001/sessions/sess-733eceff3cb946eebabe039c688d823c' \
-H 'session_id: sess-733eceff3cb946eebabe039c688d823c' \
--data-raw ''
```

Which should return:

```.json
{
    "data": {
        "session": {
            "id": "sess-733eceff3cb946eebabe039c688d823c",
            "active": true,
            "expires_at": "2020-06-04T17:10:16Z",
            "deleted": false,
            "deleted_at": null,
            "role": "Customer",
            "user_id": "cust-0694b4d4ebedf7b13a55b5734720f989",
            "user_agent": "PostmanRuntime/7.25.0"
        }
    },
    "extra": ""
}
```

ii) When session being fetched doesn't belong to user who owns the session in header
Send:

```
curl -L -X GET 'http://localhost:3001/sessions/sess-733eceff3cb946eebabe039c688d823c' \
-H 'session_id: sess-733eceff3cb946eebabe039c688d82' \
--data-raw ''
```

Which should return:

```.json
{
    "errors": [
        {
            "user": [
                "is unauthorized"
            ]
        }
    ]
}
```

c) To retrieve all sessions for a user, send:

```
curl -L -X GET 'http://localhost:3001/sessions' \
-H 'session_id: sess-733eceff3cb946eebabe039c688d823c' \
-H 'Content-Type: application/json' \
'
```

Which should return:

```.json
{
    "data": {
        "sessions": [
            {
                "id": "sess-b49b4993644c057af150db3b90adcaca",
                "active": true,
                "expires_at": "2020-06-05T09:58:47Z",
                "deleted": false,
                "deleted_at": null,
                "role": "Customer",
                "user_id": "cust-0694b4d4ebedf7b13a55b5734720f989",
                "user_agent": "Brave"
            },
            {
                "id": "sess-9d9461cfd39e827f00afb31e2490ae71",
                "active": true,
                "expires_at": "2020-06-04T10:04:18Z",
                "deleted": false,
                "deleted_at": null,
                "role": "Customer",
                "user_id": "cust-0694b4d4ebedf7b13a55b5734720f989",
                "user_agent": "PostmanRuntime/7.25.0"
            },
            {
                "id": "sess-733eceff3cb946eebabe039c688d823c",
                "active": true,
                "expires_at": "2020-06-05T17:10:16Z",
                "deleted": false,
                "deleted_at": null,
                "role": "Customer",
                "user_id": "cust-0694b4d4ebedf7b13a55b5734720f989",
                "user_agent": "PostmanRuntime/7.25.0"
            }
        ]
    },
    "extra": ""
}
```

d) To delete a session, send

```
curl -L -X DELETE 'http://localhost:3001/sessions/sess-733eceff3cb946eebabe039c688d823c' \
-H 'session_id: sess-733eceff3cb946eebabe039c688d823c' \
-H 'Content-Type: application/json' \

```

Which should return:

```.json
{
    "data": {
        "session": {
            "id": "sess-733eceff3cb946eebabe039c688d823c",
            "active": false,
            "expires_at": "2020-06-05T17:10:16Z",
            "deleted": true,
            "deleted_at": "2020-06-03T17:17:31Z",
            "role": "Customer",
            "user_id": "cust-0694b4d4ebedf7b13a55b5734720f989",
            "user_agent": "PostmanRuntime/7.25.0"
        }
    },
    "extra": ""
}
```
#### Any background context you want to provide?
N/A
### Basecamp List
https://3.basecamp.com/4328072/buckets/16770348/todolists/2717920510#adding_to_2717920510
